### PR TITLE
system-wide page caching (clock-pro based)

### PR DESCRIPTION
**Pull request by senior7515**

### CODE REVIEW

File: src/filesystem/page_cache_bucket_allocator.h [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+  inline uint64_t used_bytes() {
+    return (alloc512_.size() * kAllocatorSize)
+           + (alloc4096_.size() * kAllocatorSize);
+  }
+
+
+  void reclaim() {
+    if (used_bytes() <= max_bytes_) { return; }
+
+    auto reclaim_one = [](auto &vec) {
```
Comment (senior7515) [inline src/filesystem/page_cache_bucket_allocator.h]: need to decrease the index, but maybe we don't need it, we can always use .rbegin() ?

File: src/filesystem/page_cache_bucket_allocator.h [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    case 4096:
+      auto  a = get_alloc(alloc4096_, alloc4096_idx_, 4096);
+      retun seastar::make_lw_shared<page>(a.alloc(), &a);
+    default:
+      LOG_THROW("Could not find allocator for alignment of {}", sz);
+    }
+  }
+  inline uint64_t used_bytes() {
+    return (alloc512_.size() * kAllocatorSize)
+           + (alloc4096_.size() * kAllocatorSize);
```
Comment (hellertime) [inline src/filesystem/page_cache_bucket_allocator.h]: This is just `kAllocatorSize * (alloc512_.size() + alloc4096_.size())`

File: src/filesystem/wal_page_cache.cc [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+seastar::future<wal_page_cache::page_t> wal_page_cache::cache_read(
+  seastar::lw_shared_ptr<seastar::file> f,
+  const uint64_t &                      global_fid,
+  const uint32_t &                      original_page,
+  const seastar::io_priority_class &    pc) {
+  uint64_t                k[]      = {global_fid, original_page};
+  typename cache_t::key_t page_idx = xxhash_64((char *)&k, SMF_ARRAYSIZE(k));
+
+  DLOG_INFO("global_file_id: {}, original_page: {}, compuated page_idx: {}",
```
Comment (hellertime) [inline src/filesystem/wal_page_cache.cc]: s/compuated/computed/

File: src/filesystem/wal_page_cache.cc [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+                  page_idx, chunk.key);
+        cache_->set(std::move(chunk));
+        DLOG_INFO("PLEASE TELL ME THE TRUTH: {}", cache_->contains(page_idx));
+        return seastar::make_ready_future<page_t>(
+          cache_->get_chunk_ptr(page_idx)->data);
+      });
+  }
+  cache_->run_cold_hand();
+  cache_->run_hot_hand();
+  cache_->fix_hands();
```
Comment (hellertime) [inline src/filesystem/wal_page_cache.cc]: since this order is so important, perhaps wrap in a function? `tick_tock()`

**Comment by senior7515** [comment]
missing:

- [ ] tests
- [ ] pretty print fixes for all wal types
- [ ] finish code TODO's

**Comment by senior7515** [comment]
@noahdesu  I'm probably 1 week out (2 weeks max) from cleaning up the patch in a state that can be merged into /master. 

In its current form it works and has unit test coverage, but not enough. 

Anyway, just wanted to ping you since you have the really nice and large patch for the cmake stuff.

**Comment by senior7515** [comment]
client_comparison
server_comparison

**Comment by senior7515** [comment]
Pending PR for a flatbuffers upstream patch: https://github.com/google/flatbuffers/pull/4614

We'll merge a 'stringly' typed API w/ a TODO to refactor it later for a typesafe API.

**Pull request merged**